query/src/main/java/org/infinispan/query/concurrent/InfinispanIndexingExecutorProvider.java

```diff
@@ -0,0 +1,67 @@
+package org.infinispan.query.concurrent;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.hibernate.search.backend.lucene.work.spi.LuceneWorkExecutorProvider;
+import org.hibernate.search.engine.common.execution.spi.SimpleScheduledExecutor;
+import org.infinispan.factories.AbstractComponentFactory;
+import org.infinispan.factories.AutoInstantiableFactory;
+import org.infinispan.factories.annotations.DefaultFactoryFor;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.util.concurrent.BlockingManager;
+
+@Scope(Scopes.GLOBAL)
+@DefaultFactoryFor(classes = LuceneWorkExecutorProvider.class)
+public class InfinispanIndexingExecutorProvider extends AbstractComponentFactory implements AutoInstantiableFactory, LuceneWorkExecutorProvider {
+
+   @Inject
+   BlockingManager blockingManager;
+
+   @Override
+   public Object construct(String name) {
+      return new InfinispanIndexingExecutorProvider();
+   }
+
+   @Override
+   public SimpleScheduledExecutor writeExecutor(Context context) {
+      return new InfinispanScheduledExecutor(blockingManager);
+   }
+
+   static final class InfinispanScheduledExecutor implements SimpleScheduledExecutor {
+
+      private final BlockingManager blockingManager;
+
+      public InfinispanScheduledExecutor(BlockingManager blockingManager) {
+         this.blockingManager = blockingManager;
+      }
+
+      @Override
+      public Future<?> submit(Runnable task) {
+         return blockingManager.runBlocking(task, this).toCompletableFuture();
+      }
+
+      @Override
+      public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+         return blockingManager.scheduleRunBlocking(command, delay, unit, this);
+      }
+
+      @Override
+      public void shutdownNow() {
+         // The executors lifecycle is handled at global Infinispan level
+      }
+
+      @Override
+      public boolean isBlocking() {
+         return true;
```

Comment:
Technically this will never block if the submitting thread is a non blocking thread (would throw CacheBackpressureFullException at worst case), but may run the task in line if submitted from a different thread. Will these tasks be submitted on our Infinispan threads?

Reply:
From what I can see the submitting thread here is usually a blocking thread, for instance: `blocking-thread-SchemaIndexUpdateTest-NodeA-p3-t1`.
This thread is usually activated by our (Infinispan JavaRX) non blocking layer, which uses the blocking manager to offload the indexing blocking operations.
Going on debug I see, if don't get wrong, that the task is run in line most of the time.
From my understanding our blockingManager never blocks accepting  #runBlocking(task, this) operations and if so we can return blocking false here.
I’m going to change it.
I'm also adding a comment.

Reply:
It would never block `runBlocking` operations if invoked in a blocking thread, true. But it may run the task in the invoking thread if the blocking thread pool and its queue are completely full. But that should be fine really.